[clang] implement printing of canonical expressions

mizvekov opened the pull request:
This patch extends the canonicalization printing policy to cover expressions
and template names, and wires that up to the template argument printer,
covering expressions, and to the expression within a dependent decltype.

This is helpful for debugging, or if these expressions somehow end up
in diagnostics, as without this patch they can print as completely unrelated
expressions, which can be quite confusing.

This is because expressions are not uniqued, unlike types, and
when a template specialization containing an expression is the first to be
canonicalized, the expression ends up appearing in the canonical type of
subsequent equivalent specializations.

Fixes https://github.com/llvm/llvm-project/issues/92292

mizvekov committed:
[clang] implement printing of canonical expressions

This patch extends the canonicalization printing policy to cover expressions
and template names, and wires that up to the template argument printer,
covering expressions, and to the expression within a dependent decltype.

This is helpful for debugging, or if these expressions somehow end up
in diagnostics, as without this patch they can print as completely unrelated
expressions, which can be quite confusing.

This is because expressions are not uniqued, unlike types, and
when a template specialization containing an expression is the first to be
canonicalized, the expression ends up appearing in the canonical type of
subsequent equivalent specializations.

Fixes https://github.com/llvm/llvm-project/issues/92292

erichkeane commented on clang/lib/AST/JSONNodeDumper.cpp:
```diff
@@ -1724,6 +1724,8 @@ void JSONNodeDumper::VisitTemplateExpansionTemplateArgument(
     const TemplateArgument &TA) {
   JOS.attribute("isExpr", true);
+  if (TA.isCanonicalExpr())
+    JOS.attribute("isCanon", true);
```
Any reason to not just do `isCanonical` instead?  `Canon` is already a word and sounds nonsensical.

erichkeane commented on clang/lib/AST/StmtPrinter.cpp:
```diff
@@ -1305,9 +1305,13 @@ void StmtPrinter::VisitDeclRefExpr(DeclRefExpr *Node) {
     OS << "template ";
+
+  bool ForceAnonymous =
+      Policy.PrintAsCanonical && VD->getKind() == Decl::NonTypeTemplateParm;
```
Can you explain what is going on here?  This is a little subtle.

erichkeane commented on clang/lib/AST/TextNodeDumper.cpp:
```diff
@@ -1357,6 +1357,8 @@ void TextNodeDumper::VisitTemplateExpansionTemplateArgument(
     const TemplateArgument &TA) {
   OS << " expr";
+  if (TA.isCanonicalExpr())
+    OS << " canon";
```
Same hope here on the full word.

mizvekov replied:
Done

cor3ntin commented:
Do you have tests for cases where it would impact diagnostics?

mizvekov commented:
Sure, but I'd have to resist the urge to fix them, so that they can end up as test cases. They probably wouldn't survive long either way.

For example, I grepped the source tree for diagnostics where we are emitting 'type-parameter-X-X', and where this is happening not because the parameter is anonymous.

Found this one in `CXX/temp/temp.decls/temp.mem/p5.cpp` (reduced):
```C++
struct X0 {
  template<typename T> operator T*() const; // expected-note{{explicit instantiation refers here}}
};
template X0::operator float*() const; // expected-error{{explicit instantiation of undefined function template 'operator type-parameter-0-0 *'}}
```

Converting this test case to one involving NTTP is easy:
```C++
template<int A> struct B {};
struct X0 {
  template<int C> operator B<C>() const;
};
template X0::operator B<0>() const;
```

And sure enough, with this patch we produce:
`error: explicit instantiation of undefined function template 'operator B<value-parameter-0-0>'`
Instead of:
`error: explicit instantiation of undefined function template 'operator B<A>'`

Now you could devise a more complicated scenario where, instead of coincidentally printing the correct template parameter name, we produce an earlier equivalent template specialization with a different parameter name, and that causes a later one to be printed incorrectly.

mizvekov commented:
I've added a few test cases showing the effects on diagnostics.
Which is likely to be short lived, as the underlying issue looks easy to solve.

I have updated the patch to cover printing of dependent decltype as well.

I have looked into also covering DependentSizedArrays, but the sugaring there is broken at the moment
and would need a separate and more involved fix before we can turn this on there.
